Solutions/week2/e02

adamJacewicz opened the pull request.

michaljabi committed:
init: week3 tasks - good luck !

michaljabi committed:
init: week4 tasks - good luck !

adamJacewicz committed:
s

adamJacewicz committed:
- added button that create new order
- use rxjs to refresh data

adamJacewicz committed:
- refactor

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+  orderNumber,
+});
+
+export class Orders {
```
bardzo ładnie zafasadowana logika dla `Orders` razem ze strumieniami.

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  getBestSale(): Observable<number> {
+    return this.#source.pipe(map(getBestSale));
```
pamiętaj że to można było już zrobić w polu klasy np:
```typescript
#bestSales = this.#source.pipe(map(getBestSale));
```
a potem tutaj:

```typescript
getBestSale(): Observable<number> {
  return this.#bestSales;
}
```

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+
+  async getAllOrders(): Promise<Observable<OrderType[]>> {
+    await wait(1000);
+    return this.#source.asObservable();
```
Nie rozumiem dlaczego tutaj `Promise` zwracasz. można by było normalnie `Observable<OderType[]>`

michaljabi replied:
Dodatkowo nie potrzebujesz czegoś takiego jak `wait(1000)`
RxJs sam sobie poradzi z takimi tematami, zobacz `timer` oraz `switchMap`
https://www.learnrxjs.io/learn-rxjs/operators/creation/timer

https://www.learnrxjs.io/learn-rxjs/operators/transformation/switchmap

```typescript
getAllOrders(): Observable<OrderType[]> {
    return timer(1000).pipe(switchMap(() => this.#source))
}
```

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+      `${this.#source.value.length}/${new Date().getFullYear()}`
+    );
+    this.#source.next([...this.#source.value, newOrder]);
+  }
```
tak samo:
```typescript
timer(1000).subscribe(() => {
 const newOrder = createOrder(
      sale,
      `${this.#source.value.length}/${new Date().getFullYear()}`
    );
    this.#source.next([...this.#source.value, newOrder]);
})
```

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/orders/helpers.ts:
```diff
@@ -0,0 +1,36 @@
+      Number(c) ^
+      (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (Number(c) / 4)))
+    ).toString(16)
+  );
```
fajny generator

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  getTotalIncome(): Observable<number> {
+    return this.#source.pipe(map(getTotalIncome));
```
bardzo ładnie zrealizowany temat `helpers` gdzie pozamykałeś @adamJacewicz te funkcyjki sobie i mają swoje miejsce.

Ja bym jedynie dał inną nazwe na ich mieszkanie `helpers.ts` to dla mnie zbyt "generyczna" nazwa. nie mówiąca zbyt wiele.

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/main.ts:
```diff
@@ -0,0 +1,20 @@
+  setTotalIncome(ordersStore.getTotalIncome());
+  setTotalOrders(ordersList);
+  initAddOrderButton(ordersStore);
+};
```
To rozwiązanie najmniej mi się podoba z całego kodu (reszta koncepcyjnie jest bardzo w porządku, fajowo że korzystasz z `RxJS` i wiesz jak sobie `Ramde` wpiąć we własne funkcje i zastosować kompozycję.

Tutaj `init` robi za dużo.
Powinien zainicjować każdy komponent ale dać mu pełny `ordersStore`
i wtedy każdy komponent SAM zdecyduje do jakiego strumienia ordersStore się wpina i subskrybuje zmiany.

Tutaj `init` podejmuje niepotrzebnie za dużo decycji o tym kto co dostanie.
Komponent sam sobie to powinien "ogarniać"

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/model/orders.ts:
```diff
@@ -0,0 +1,56 @@
+  wait,
+} from "../orders/helpers";
+
+export type OrderType = {
```
To bym dał do osobnego pliku w tym folderze.

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/orders/helpers.ts:
```diff
@@ -0,0 +1,36 @@
+export const sum = (list) => list.reduce((a, b) => a + b, 0).toFixed(2);
+export const getSales = (list) => list.map(getProperty("sale"));
+export const getBestSale = (list) => compose(getHighestValue, getSales)(list);
+export const getTotalIncome = (list) => compose(sum, getSales)(list);
```
tutaj te parametry `list` są implicit `:any`

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/orders/tiles/AnnualOrdersTile.ts:
```diff
@@ -0,0 +1,30 @@
+  const componentId = "annual-orders";
+  const mountPoint = document.querySelector(`[data-tile="${componentId}"]`);
+  const subTitle = mountPoint?.querySelector("[data-subtitle]");
+  const panel = mountPoint?.querySelector("[data-panel]");
```
te `const` mogą byc poza tą funkcją.

michaljabi commented on week-2-wide-perspective/w02e02-data-manipulation/src/orders/tiles/AddOrderButton.ts:
```diff
@@ -0,0 +1,10 @@
+import { Orders } from "../../model/orders";
+
+export const initAddOrderButton = (store: Orders) => {
+  const button = document.querySelector('[data-button="add-order"]');
```
selektor daj poza funkcją. będzie wydajniej.

michaljabi reviewed (changes requested):
Uwaga ogólna:

- trochę dużo arrow functions :D 
- w `helperach` to jeszcze ujdzie - rozumiem
- ale w np,. komponentach czy `init` - to mogą byc zwykłe `functions`

nie jest to błąd oczywiście, ale generalnie `arrow` jest tam gdzie `callbacks` oraz tam gdzie nie da się inaczej + ew. helpers (one-liners)

natomiast tu i tam w kodzie czasem warto skorzystać ze zjawiska hoistingu funkcji - tam gdzie to bezpieczne.
no i z zapisu `function` dla samej wizualnej organizacji kodu.

---

Super że użyłeś `RxJS` ! generalnie fajnie to wyszło.

adamJacewicz committed:
- used `timer` method provided by rxjs instead of custom async approach
- used class properties over methods where it is possible
- moved order type to separated file
- corrected function argument names
- const variables moved outside methods where it is possible
- refactor

adamJacewicz commented:
> Uwaga ogólna:
> 
> * trochę dużo arrow functions :D
> * w `helperach` to jeszcze ujdzie - rozumiem
> * ale w np,. komponentach czy `init` - to mogą byc zwykłe `functions`
> 
> nie jest to błąd oczywiście, ale generalnie `arrow` jest tam gdzie `callbacks` oraz tam gdzie nie da się inaczej + ew. helpers (one-liners)
> 
> natomiast tu i tam w kodzie czasem warto skorzystać ze zjawiska hoistingu funkcji - tam gdzie to bezpieczne. no i z zapisu `function` dla samej wizualnej organizacji kodu.
> 
> Super że użyłeś `RxJS` ! generalnie fajnie to wyszło.


Dzięki za wskazówki, kod poprawiony + zrobiłem mały refactor

adamJacewicz committed:
- use rxjs event listening approach

michaljabi reviewed (approved):
Bardzo ładne poprawki - brawo 🥇
